File: 0402-remove-k-digits/0402-remove-k-digits.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        vector<char> stack;

        for (char digit : num) {
            while (!stack.empty() && k > 0 && stack.back() > digit) {
                stack.pop_back();
                k--;
            }
            stack.push_back(digit);
        }

        // Remove remaining digits from the end if k > 0
        while (k > 0 && !stack.empty()) {
            stack.pop_back();
            k--;
        }

        // Build the result string and remove leading zeros
        string result = "";
        bool leadingZero = true;

        for (char digit : stack) {
            if (leadingZero && digit == '0')
                continue;
            leadingZero = false;
            result += digit;
        }

        return result.empty() ? "0" : result;
    }
};

File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        
        Stack<Character> stack = new Stack<>();
        
        for(char ch : num.toCharArray()){
            while(!stack.isEmpty() && stack.peek()>ch && k>0){
                stack.pop();
                k--;
            }
            stack.push(ch);
        }

        while(k>0 && !stack.isEmpty()){
            stack.pop();
            k--;
        }

        String result = "";
        while(!stack.isEmpty()){
            result= stack.pop()+ result;
        }
        int index = 0;
        while (index < result.length() && result.charAt(index) == '0') {
            index++;
        }
        result = result.substring(index);
        if(result == ""){
            return "0";
        }
        return result;
    }
}